File: server/smart-cache.ts
import type { Pool } from "pg";
import type { Smart } from "@shared/schema";
import { normalizeArticle } from "@shared/normalization";

type SmartInternal = Smart & {
  normalizedSmart: string;
  normalizedArticles: string[];
};

export type SmartCache = {
  size: number;
  getBySmart: (smart: string) => Smart | undefined;
  search: (normalizedQuery: string, limit?: number) => Smart[];
};

function toArray(value: unknown): string[] {
  if (!value) return [];
  if (Array.isArray(value)) return value.filter((v) => typeof v === "string");
  if (typeof value === "string") return [value];
  return [];
}

export async function loadSmartCache(partsPool: Pool): Promise<SmartCache> {
  const res = await partsPool.query<{
    smart: string;
    articles: unknown;
    name: string | null;
    brand: unknown;
    description: unknown;
  }>(`
    SELECT
      smart,
      "артикул" as articles,
      "наименование" as name,
      "бренд" as brand,
      "коннект_бренд" as description
    FROM public.smart
  `);

  const items: SmartInternal[] = res.rows.map((row) => {
    const articles = toArray(row.articles);
    return {
      smart: row.smart,
      articles,
      name: row.name,
      brand: toArray(row.brand),
      description: toArray(row.description),
      normalizedSmart: normalizeArticle(row.smart),
      normalizedArticles: articles.map((a) => normalizeArticle(a)),
    };
  });

  const bySmart = new Map<string, SmartInternal>();
  for (const item of items) {
    bySmart.set(item.smart, item);
  }

  return {
    size: items.length,
    getBySmart: (smartCode: string) => bySmart.get(smartCode),
    search: (normalizedQuery: string, limit = 50) => {
      const q = normalizeArticle(normalizedQuery);
      if (!q) return [];

      const scored: Array<{ score: number; item: SmartInternal }> = [];

      for (const item of items) {
        let best = 0;

        if (item.normalizedSmart === q) best = Math.max(best, 70);
        else if (item.normalizedSmart.startsWith(q)) best = Math.max(best, 60);
        else if (item.normalizedSmart.includes(q)) best = Math.max(best, 50);

        for (const art of item.normalizedArticles) {
          if (art === q) best = Math.max(best, 100);
          else if (art.startsWith(q)) best = Math.max(best, 90);
          else if (art.includes(q)) best = Math.max(best, 80);
        }

        if (best > 0) scored.push({ score: best, item });
      }

      scored.sort((a, b) => b.score - a.score || a.item.smart.localeCompare(b.item.smart));

      return scored.slice(0, limit).map(({ item }) => ({
        smart: item.smart,
        articles: item.articles,
        name: item.name,
        brand: item.brand,
        description: item.description,
      }));
    },
  };
}

File: py_backend/smart_cache.py
"""SMART cache loading and in-memory search."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Iterable

import asyncpg
from pydantic import BaseModel, Field

from .normalization import normalize_article


class Smart(BaseModel):
    smart: str
    articles: list[str] = Field(default_factory=list)
    name: str | None = None
    brand: list[str] = Field(default_factory=list)
    description: list[str] = Field(default_factory=list)


@dataclass
class _SmartInternal:
    smart: str
    articles: list[str]
    name: str | None
    brand: list[str]
    description: list[str]
    normalized_smart: str
    normalized_articles: list[str]

    def as_public(self) -> Smart:
        return Smart(
            smart=self.smart,
            articles=list(self.articles),
            name=self.name,
            brand=list(self.brand),
            description=list(self.description),
        )


def to_array(value: Any) -> list[str]:
    if not value:
        return []
    if isinstance(value, list):
        return [item for item in value if isinstance(item, str)]
    if isinstance(value, str):
        return [value]
    return []


class SmartCache:
    def __init__(self, items: Iterable[_SmartInternal]):
        self._items = list(items)
        self._by_smart = {item.smart: item for item in self._items}

    @property
    def size(self) -> int:
        return len(self._items)

    def get_by_smart(self, smart: str) -> Smart | None:
        item = self._by_smart.get(smart)
        return item.as_public() if item else None

    # TypeScript-compatible name used by the ported storage layer.
    def getBySmart(self, smart: str) -> Smart | None:
        return self.get_by_smart(smart)

    def search(self, normalized_query: str, limit: int = 50) -> list[Smart]:
        q = normalize_article(normalized_query)
        if not q:
            return []

        scored: list[tuple[int, _SmartInternal]] = []
        for item in self._items:
            best = 0

            if item.normalized_smart == q:
                best = max(best, 70)
            elif item.normalized_smart.startswith(q):
                best = max(best, 60)
            elif q in item.normalized_smart:
                best = max(best, 50)

            for article in item.normalized_articles:
                if article == q:
                    best = max(best, 100)
                elif article.startswith(q):
                    best = max(best, 90)
                elif q in article:
                    best = max(best, 80)

            if best > 0:
                scored.append((best, item))

        scored.sort(key=lambda pair: (-pair[0], pair[1].smart))
        return [item.as_public() for _, item in scored[:limit]]


async def load_smart_cache(parts_pool: asyncpg.Pool) -> SmartCache:
    rows = await parts_pool.fetch(
        """
        SELECT
          smart,
          "артикул" as articles,
          "наименование" as name,
          "бренд" as brand,
          "коннект_бренд" as description
        FROM public.smart
        """
    )

    items: list[_SmartInternal] = []
    for row in rows:
        articles = to_array(row["articles"])
        smart = str(row["smart"])
        items.append(
            _SmartInternal(
                smart=smart,
                articles=articles,
                name=row["name"],
                brand=to_array(row["brand"]),
                description=to_array(row["description"]),
                normalized_smart=normalize_article(smart),
                normalized_articles=[normalize_article(article) for article in articles],
            )
        )

    return SmartCache(items)


# TypeScript-compatible aliases.
loadSmartCache = load_smart_cache
